sky/skylet/providers/gcp/config.py

```diff
@@ -722,6 +755,84 @@ def _merge_and_refine_rule(rules):
     return True
 
 
+def _get_filewall_rules_template(config):
+    """Returns the firewall rules template."""
+    default_template = FIREWALL_RULES_TEMPLATE.copy()
+    ports = config["provider"].get("ports", [])
+    for port in ports:
+        suffix = f"-user-ports-{config['cluster_name']}-{port}"
+        default_template.append(
+            {
+                "name": "{VPC_NAME}" + suffix,
+                "description": f"Allow user-specified port {port} for cluster {config['cluster_name']}",
+                "network": "projects/{PROJ_ID}/global/networks/{VPC_NAME}",
+                "selfLink": "projects/{PROJ_ID}/global/firewalls/{VPC_NAME}" + suffix,
+                "direction": "INGRESS",
+                "priority": 65534,
+                "allowed": [
+                    {
+                        "IPProtocol": "tcp",
+                        "ports": [str(port)],
+                    },
+                ],
+                "sourceRanges": ["0.0.0.0/0"],
+                "targetTags": [config["cluster_name"]],
+            }
+        )
+    return default_template
+
+
+def _is_same_rule(template, existing, PROJ_ID, VPC_NAME):
+    # TODO(tian): simplify this function
```

Comment:
Yes, we need to simplify the function or at least add some explanation in comment for what this is going to do.